pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,230 @@
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	rbacv1informers "k8s.io/client-go/informers/rbac/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// RequeueDelay is time to wait before requeue the key
+var RequeueDelay = 2 * time.Second
+
+type listManifestWorksFunc func() ([]*workapiv1.ManifestWork, error)
+
+// FinalizeController ensures all manifestworks are deleted before role/rolebinding for work
+// agent are deleted in a terminating cluster namespace.
+type FinalizeController struct {
+	roleLister         rbacv1listers.RoleLister
+	roleBindingLister  rbacv1listers.RoleBindingLister
+	rbacClient         rbacv1client.RbacV1Interface
+	namespaceLister    corelisters.NamespaceLister
+	manifestWorkClient workv1client.WorkV1Interface
+	manifestWorkLister worklister.ManifestWorkLister
+	eventRecorder      events.Recorder
+}
+
+func NewFinalizeController(
+	eventRecorder events.Recorder,
+	roleInformer rbacv1informers.RoleInformer,
+	roleBindingInformer rbacv1informers.RoleBindingInformer,
+	namespaceLister corelisters.NamespaceLister,
+	manifestWorkClient workv1client.WorkV1Interface,
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+) factory.Controller {
+
+	controller := &FinalizeController{
+		roleLister:         roleInformer.Lister(),
+		roleBindingLister:  roleBindingInformer.Lister(),
+		namespaceLister:    namespaceLister,
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		rbacClient:         rbacClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("FinalizeController", eventRecorder)
+}
+
+func (m *FinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+
+	// ignore role/rolebinding whose name does not end with the expected suffix
+	if !strings.HasSuffix(key, keySuffix) {
+		return nil
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+	clusterName := strings.TrimSuffix(name, keySuffix)
+
+	// ignore role/rolebinding if namespace does not match with cluster name
+	if namespace != clusterName {
+		return nil
+	}
+
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
```

Comment:
proposal

```go
if !ns.terminating{
    // unconditionally remove the finalizer
    return
}

numManifestWorksRemaining :=  m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())
if numManifestWorksRemaining > 0{
    queue.AddRateLimited(key)
    return nil
}

if objectMetaHasFinalizer(role){
    remove finalizer
}

if objectMetaHasFinalizer(rolebinding){
    remove finalizer
}
```

This has minimal nesting and early returns which made the code very easy to reason about.